lib/AST/Type.cpp

```diff
@@ -2941,6 +2962,8 @@ LookUpConformanceInSubstitutionMap::operator()(CanType dependentType,
   if (!conformingReplacementType)
     return None;
 
+  // TODO: Describe what this code is doing. It works around the issues
+  // in the conformance lookup, when a dependent type cannot be found...
```

Comment:
It sounds like either you're not adding the right conformances when building a SubstitutionMap, or SubstitutionMap::lookupConformance() is not returning the right thing.

I'm planning on re-doing SubstitutionMap construction so that it is driven by callbacks rather than calls to addSubstitution()/addConformance() on the map. This will ensure the correct entries are always added.